### PR TITLE
Fix watch for local registry

### LINKED ISSUE #217: Watch for CAS service accounts in local registry stucks

### PR DESCRIPTION
**Pull request by nhinze23**
Resolves #217

### CODE REVIEW

File: app/src/main/java/de/triology/cas/services/RegistryLocal.java [hunk 1]
```diff
@@ -217,31 +222,73 @@ public URI getCasLogoutUri(String doguname) throws GetCasLogoutUriException {
...
                     }
                 }
-            } catch (IOException e) {
-                throw new RegistryException("Failed to addDoguChangeListener", e);
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new RegistryException("Failed to addDoguChangeListener", e);
+
+                if (!serviceAccountRegistryChanged) {
+                    continue;
```
Comment (meiserloh) [inline app/src/main/java/de/triology/cas/services/RegistryLocal.java:255]: If I understand the comment below correctly we need to reset the key in the `continue`-cases too, don't we?

File: app/src/main/java/de/triology/cas/services/RegistryLocal.java [hunk 2]
```diff
@@ -217,31 +222,73 @@ public URI getCasLogoutUri(String doguname) throws GetCasLogoutUriException {
...
+                var currentServiceAccounts = readServiceAccounts();
+                if (!previousServiceAccounts.deepEquals(currentServiceAccounts)) {
+                    LOGGER.info("services changed. call doguChangeListener");
+                    doguChangeListener.onChange();
+                    previousServiceAccounts = currentServiceAccounts;
+                }
+
+                // Resetting the watchKey is very important. Otherwise, the key returned from take() (or poll()) will not return any more events.
+                if (!key.reset()) {
+                    LOGGER.info("reset key is no longer valid");
```
Comment (meiserloh) [inline app/src/main/java/de/triology/cas/services/RegistryLocal.java:267]: reset key sounds like a key that is used for resetting, not a key that is being reset. I think "watch key" or something would fit better

File: app/src/main/java/de/triology/cas/services/RegistryLocal.java [hunk 3]
```diff
@@ -217,31 +222,73 @@ public URI getCasLogoutUri(String doguname) throws GetCasLogoutUriException {
...
+    }
+
+    private WatchService initializeWatchService(Path path) throws IOException {
+        WatchService watchService = fileSystem.newWatchService();
+        registerServiceRegistryWatch(path, watchService);
+
+        return watchService;
+    }
+
+    private void registerServiceRegistryWatch(Path path, WatchService watchService) throws IOException {
```
Comment (meiserloh) [inline app/src/main/java/de/triology/cas/services/RegistryLocal.java:287]: I don't really see the point of this method (eventhough I call myself a fan of small methods). It's just a void-one liner that is called once. It's subjective, so you can decide, but I would inline this method.

File: app/src/test/java/de/triology/cas/services/RegistryLocalTest.java [hunk 4]
```diff
@@ -502,6 +506,77 @@ public void addDoguChangeListener() throws IOException, InterruptedException {
...
+                .thenReturn(watchService)
+                .thenReturn(watchService)
+                .thenReturn(watchService);
+
+        var registry = spy(RegistryLocal.class);
+        registry.fileSystem = fileSystem;
+        var initialServiceAccounts = new RegistryLocal.ServiceAccounts();
+        var changedServiceAccounts = new RegistryLocal.ServiceAccounts();
+        changedServiceAccounts.setCas(Map.of("usermgt", new RegistryLocal.ServiceAccountCas()));
+        var unchangedServiceAccounts2 = new RegistryLocal.ServiceAccounts();
```
Comment (meiserloh) [inline app/src/test/java/de/triology/cas/services/RegistryLocalTest.java:544]: Why the 2? there is no other unchangedServiceAccounts-variable

File: app/src/test/java/de/triology/cas/services/RegistryLocalTest.java [hunk 5]
```diff
@@ -502,6 +506,77 @@ public void addDoguChangeListener() throws IOException, InterruptedException {
...
+                dogus.add("dogu " + dogus.size());
+                dogus.notify();
+            }
+        });
+
+        synchronized (dogus) {
+            dogus.wait();
+        }
+
+        assertThat(dogus, hasSize(1));
```
Comment (meiserloh) [inline app/src/test/java/de/triology/cas/services/RegistryLocalTest.java:565]: I don't get this assertion. What is it checking? That the Listener is not changing the amount of dogus?

Comment (nhinze23) [reply app/src/test/java/de/triology/cas/services/RegistryLocalTest.java:565]: Yes. Just a change after a reinitialisation.

File: app/src/main/java/de/triology/cas/services/RegistryLocal.java (hunk 1 above)
Comment (nhinze23) [reply app/src/main/java/de/triology/cas/services/RegistryLocal.java:255]: Nice catch :+1:

**Pull request merged**